rust/parquet/src/record/display.rs

```diff
@@ -0,0 +1,164 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Utility structs and methods to help with displaying schemas
+
+use std::fmt::{self, Display, Write};
+
+use super::Schema;
+use crate::basic::{LogicalType, Repetition};
+
+/// Implement [`Display`] given a closure that accepts a [`fmt::Formatter`] and returns a
+/// [`fmt::Result`].
+pub struct DisplayFmt<F>(F)
+where
+    F: Fn(&mut fmt::Formatter) -> fmt::Result;
+impl<F> DisplayFmt<F>
+where
+    F: Fn(&mut fmt::Formatter) -> fmt::Result,
+{
+    pub fn new(f: F) -> Self {
+        Self(f)
+    }
+}
+impl<F> Display for DisplayFmt<F>
+where
+    F: Fn(&mut fmt::Formatter) -> fmt::Result,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0(f)
+    }
+}
+
+// The following is code adapted from https://github.com/rust-lang/rust/blob/79d8a0fcefa5134db2a94739b1d18daa01fc6e9f/src/libcore/fmt/builders.rs
+
+struct PadAdapter<T: fmt::Write> {
+    fmt: T,
+    on_newline: bool,
+}
+
+impl<T> PadAdapter<T>
+where
+    T: fmt::Write,
+{
+    fn new(fmt: T) -> Self {
+        PadAdapter {
+            fmt,
+            on_newline: false,
+        }
+    }
+}
+
+impl<T> fmt::Write for PadAdapter<T>
+where
+    T: fmt::Write,
+{
+    fn write_str(&mut self, mut s: &str) -> fmt::Result {
+        while !s.is_empty() {
+            if self.on_newline {
+                self.fmt.write_str("    ")?;
+            }
+
+            let split = match s.find('\n') {
+                Some(pos) => {
+                    self.on_newline = true;
+                    pos + 1
+                }
+                None => {
+                    self.on_newline = false;
+                    s.len()
+                }
+            };
+            self.fmt.write_str(&s[..split])?;
+            s = &s[split..];
+        }
+
+        Ok(())
+    }
+}
+
+/// Display a group. Tuples, [`Group`](super::types::Group) and structs marked with
+/// `#[derive(Record)]` all make use of this, so the logic is encapsulated here.
+#[must_use = "must eventually call `finish()` on Debug builders"]
```

Comment:
I haven't tested this specifically but I don't doubt it works as specified. This code is taken from the rust standard library (see the link just above this in the code) but as this struct isn't supposed to be used by users, only internally and by the derive macro, then I have no issue removing it.

Reply:
I think it only check whether the returned result is used or not so in this case it may not check whether the method is called or not. If not useful, please remove.

Reply:
Ah I see your concern. `must_use` does work on types as well as return values (see the [reference](https://doc.rust-lang.org/reference/attributes.html#must_use)) so I'll leave this in.